rpcs3/Emu/Cell/lv2/sys_usbd.cpp

```diff
@@ -1,219 +1,559 @@
-#include "stdafx.h"
+﻿#include "stdafx.h"
 #include "sys_usbd.h"
 
+#include <queue>
+#include <thread>
+#include "Emu/Memory/vm.h"
 #include "Emu/System.h"
 
+#include "Emu/Cell/PPUThread.h"
 #include "Emu/Cell/ErrorCodes.h"
 #include "sys_ppu_thread.h"
 
+#include "Emu/Io/Skylander.h"
+
+#include <libusb.h>
 
 LOG_CHANNEL(sys_usbd);
 
-std::vector<usbDevice> devices = {
-	// System devices
-	usbDevice{ DeviceListUnknownDataType{0x1, 0x2, 0x2, 0x44}, 50, deviceDescriptor{0x12, 0x1, 0x0200, 0, 0, 0,			0x40, 0x054C, 0x0250, 0x0009, 3, 4, 5, 1}},
-	usbDevice{ DeviceListUnknownDataType{0x1, 0x3, 0x2, 0xAD}, 326, deviceDescriptor{0x12, 0x1, 0x0200, 0xE0, 0x1, 0x1,	0x40, 0x054C, 0x0267, 0x0001, 1, 2, 0, 1}},
-	// USB Drive
-	usbDevice{ DeviceListUnknownDataType{0x0, 0x4, 0x2, 0x0}, 50, deviceDescriptor{0x12, 0x1, 0x0200, 0, 0, 0,			0x40, 0x1516, 0x1226, 0x0100, 1, 2, 3, 1}},
-	// Skylanders Portal
-	usbDevice{ DeviceListUnknownDataType{0x0, 0x5, 0x2, 0xF8}, 59, deviceDescriptor{0x12, 0x1, 0x0200, 0, 0, 0,			0x20, 0x1430, 0x0150, 0x0100, 1, 2, 0, 1}},
-};
-
-bool has_register_extra_ldd;
-int receive_event_called_count;
-
-/*
- * sys_usbd_initialize changes handle to a semi-unique identifier.
- * identifier generation is speculated to be:
- * f(n+1) = (f(n) + 2<<9) %  14847 + f_min (probably just f(0)-14847 but it could be different)
- * Thanks to @flash-fire for thinking up the identifier generation code.
- * TODO: try to get hardware to return not CELL_OK (perhaps set handle to NULL).
- */
+template <>
+void fmt_class_string<libusb_transfer>::format(std::string& out, u64 arg)
+{
+	const auto& transfer = get_object(arg);
+
+	std::string datrace;
+	const char hex[16] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+
+	for (int index = 0; index < transfer.actual_length; index++)
+	{
+		datrace += hex[transfer.buffer[index] >> 4];
+		datrace += hex[(transfer.buffer[index]) & 15];
+		datrace += ' ';
+	}
+
+	fmt::append(out, "TR[r:%d][sz:%d] => %s", (u8)transfer.status, transfer.actual_length, datrace);
+}
+
+void LIBUSB_CALL callback_transfer(struct libusb_transfer* transfer)
+{
+	auto usbh = g_idm->lock<named_thread<usb_handler_thread>>(0);
+	if (!usbh)
+		return;
+
+	usbh->transfer_complete(transfer);
+}
+
+usb_handler_thread::usb_handler_thread()
+{
+	if (libusb_init(&ctx) < 0)
+		return;
+
+	// look if any device which we could be interested in is actually connected
+	libusb_device** list;
+	ssize_t ndev = libusb_get_device_list(ctx, &list);
+
+	bool found_skylander = false;
+
+	for (ssize_t index = 0; index < ndev; index++)
+	{
+		libusb_device_descriptor desc;
+		libusb_get_device_descriptor(list[index], &desc);
+
+		auto check_device = [&](const u16 id_vendor, const u16 id_product_min, const u16 id_product_max, const char* s_name) -> bool {
```

Comment:
newline

Reply:
this & all the other newline are results of our clang format, not changing each of those manually.
I'll see if I can update the clang format maybe.

Reply:
I looked recently but that flag is still not implemented